### PR TITLE
Add env variable to override sse2 optimizations and improve sse2 autodetection

### PR DESCRIPTION
**Pull request by jameshilliard**
This is useful for cross compiling since platform.machine() is broken for cross builds.

### CODE REVIEW

**Comment by jameshilliard** [comment]
I've updated this so that the auto-detection should generally work when cross compiling, I've left the env flag for handling edge-cases that might not work still.

**Comment by hynek** [comment]
That's great but invasive stuff – any ideas how to test that?

**Comment by jameshilliard** [comment]
@hynek I'm using [buildroot](https://buildroot.org/) for cross compilation to test it. But I'd say just try and regression test it on your systems to make sure nothing broke, other packages use a similar autodetection technique so I think this should be fairly safe/unlikely to cause issues.

**Comment by jameshilliard** [comment]
Once this is merged and the buildroot package is updated buildroot has an autobuilder CI system that tests packages against many different configurations/toolchains.

**Comment by jameshilliard** [comment]
@hynek Does this look ok to merge?

File: setup.py [hunk 1]
```diff
@@ -19,9 +19,31 @@
...
-# Optimized version requires SSE2 extensions.  They have been around since
-# 2001 so we try to compile it on every recent-ish x86.
-optimized = platform.machine() in ("i686", "x86", "x86_64", "AMD64")
+
+def has_sse_flag():
+    """Return a boolean indicating whether the sse flag name is supported on
+    the default compiler.
+    """
+    import tempfile
+    from distutils.ccompiler import new_compiler
```
Comment (hynek) [inline setup.py:28]: Any particular reason, why these imports are hidden away in here?

File: setup.py [hunk 2]
```diff
@@ -19,9 +19,31 @@
...
+
+    compiler = new_compiler()
+    with tempfile.NamedTemporaryFile("w", suffix=".cpp") as f:
```
Comment (hynek) [inline setup.py:31]: `.c` should be fine, no? argon2 itself is just C too.

File: setup.py [hunk 3]
```diff
@@ -19,9 +19,31 @@
 NAME = "argon2-cffi"
 PACKAGES = find_packages(where="src")
 
-# Optimized version requires SSE2 extensions.  They have been around since
-# 2001 so we try to compile it on every recent-ish x86.
-optimized = platform.machine() in ("i686", "x86", "x86_64", "AMD64")
+
+def has_sse_flag():
+    """Return a boolean indicating whether the sse flag name is supported on
```
Comment (hynek) [inline setup.py:24]: Newline after `"""` please.

File: setup.py (hunk 1 above)
Comment (jameshilliard) [reply setup.py:28]: Mostly because this codepath that needs those imports gets bypassed if the env variable is used.

File: setup.py (hunk 2 above)
Comment (jameshilliard) [reply setup.py:31]: yeah, should be fine, I'll change that

**Comment by jameshilliard** [comment]
> But it needs documentation:
> 
>     * An entry to CHANGELOG.rst
> 
>     * An explanation to docs/installations.rst. Either before or after "Using a System-wide Installation of Argon2"

Added.

**Comment by jameshilliard** [comment]
I fixed a bug where this wasn't working under some circumstances due to missing `customize_compiler(compiler)`. I've currently tested this on the following platforms and it's now working as expected, native OSX x86_64(sse2 enabled), native Linux x86_64(sse2 enabled), cross Linux arm64(sse2 disabled). If you are set up to test under windows that may be a good idea as I don't have a dev env for that at the moment.

**Comment by hynek** [comment]
There seems to be two problems on Windows (I also have no Windows, but our CI has fortunately):

https://github.com/hynek/argon2-cffi/pull/59/checks?check_run_id=639415726#step:4:7

1. it doesn't know about the option (which makes sense I guess? it's not a MSVC option)
2. `c1: fatal error C1083: Cannot open source file: 'C:\Users\RUNNER~1\AppData\Local\Temp\tmpnspvsah6.c': Permission denied` which is a lot more headscratchy.

---

But now that I've looked at the code…wouldn't it make more sense to try to compile opt.c and detect using that, while allowing the override?

IOW: try to compile it and if it works, use it. If it doesn't: don't?

Random data point: `-msse2` is a NOP on amd64 anyway.

**Comment by jameshilliard** [comment]
> which is a lot more headscratchy

Yeah, maybe a weird VM thing? Anyways I think I'll maybe just disable this autodetection on windows.

> But now that I've looked at the code…wouldn't it make more sense to try to compile opt.c and detect using that, while allowing the override?

Well I think that might be tricky due to it being build as part of other sources.

> IOW: try to compile it and if it works, use it. If it doesn't: don't?

This sounds somewhat problematic to me, I mean since the compilation run has multiple sources backing out and retrying with the setting changed doesn't seem like a good idea.

**Comment by jameshilliard** [comment]
So I've gone ahead and essentially restored the previous behavior for windows systems.

**Comment by hynek** [comment]
I guess the idea would be to take some minimal source file that uses SSE2?  Probably something with https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/tmmintrin.h or so? 🤔

**Comment by jameshilliard** [comment]
> I guess the idea would be to take some minimal source file that uses SSE2? Probably something with https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/tmmintrin.h or so?

Well the source shouldn't affect the compilation failing/not failing, we're really testing if the compiler flag works not if a particular source file compiles.

**Comment by jameshilliard** [comment]
Basically the reason you really don't want to be using anything other than a dummy program like this for test compiles is that you would make your build brittle since compilation failure for a non-dummy program could occur for many reasons other than the missing compiler flag.

**Comment by hynek** [comment]
The problem I've got with that is that we're cancelling out the compiler abstraction as we've seen with the Windows example.

Circling back to your own, original requirement: how about if we leave the detection code as it is, but allow the override via env variable? The argument you're using for non-dummy programs is something we've seen in CI for the dummy one too.

My goal here is to enable your use case without any collateral damage, if that makes sense?

**Comment by jameshilliard** [comment]
> The problem I've got with that is that we're cancelling out the compiler abstraction as we've seen with the Windows example.

Well it looks like we already do that to some degree as we don't pass the flag for windows anyways, so I think we're ok here.

> how about if we leave the detection code as it is, but allow the override via env variable?

Well it probably wouldn't be as reliable outside of buildroot but that would be better than what we have right now.

> My goal here is to enable your use case without any collateral damage, if that makes sense?

I think in general on unix like systems this is technique is fairly safe since it's used in other packages.

**Comment by jameshilliard** [comment]
> how about if we leave the detection code as it is, but allow the override via env variable?

I've submitted the original version of this PR in #61 which did just that.

**Comment by hynek** [comment]
Closing in favor of #61 – thanks for being a great sport!
